### ISSUE TITLE
ALSA configuration best practices

### ISSUE THREAD

**Issue by mmoduu** [bondagit/aes67-linux-daemon#107]
Hi!

I'm struggling with ALSA configuration. Can someone write some basic examples of how to make a playback device, or multiple devices, that are then visible to applications (like Reaper) for usage, and capture devices. Earlier, I tried a playback device but had a humongous buffer (1 sec or more) which I wasn't able to make smaller.

I'm just a casual user that likes to use AoIP at home, like listening to Spotify or watch YouTube videos. And sometimes open up a DAW and record a speech or something. Also use Discord. 

I have a Ravenna-enabled device that has some mic inputs and headphone outputs.

**Comment by bondagit** [comment]
any update on this ?

**Comment by mmoduu** [comment]
Hi!

Thanks for the example! I was able to make it work in some way, but it sure is a bit quirky to set up. And it seems that Pipewire is now paving the road, and many apps are starting to support it only. 

I haven't been able to get REAPER to work with this driver.

Spotify is problematic too.

What would be the best way to choose one of the sources to be a system default?

**Comment by mmoduu** [comment]
Okay, I just made progress! It seems that Spotify was just so picky about the configuration. I have yet to test with REAPER...

This is what I have now, hope someone will find this useful, I'm playing Spotify and can simultaneously play Youtube video from Brave browser, all sounding like playing in a correct speed:

```
pcm.source_1 {
        type dmix
        ipc_key 0x11111
        slave {
                pcm "hw:CARD=RAVENNA"
                buffer_size 1024
                channels 4
                rate 48000
                format S24_3LE
        }
        bindings { 0 0 1 1 }
}

pcm.source_2 {
        type dmix
        ipc_key 0x11111
        slave {
                pcm "hw:CARD=RAVENNA"
                buffer_size 1024
                channels 4
                rate 48000
                format S24_3LE
        }
        bindings { 0 2 1 3 }
}

pcm.!default {
        type plug
        slave.pcm source_1
}
```

**Comment by mmoduu** [comment]
False alarm, I noticed that Brave is outputting garbled audio, while Spotify is clean. Weird....

**Comment by mmoduu** [comment]
It seems that the problem lies within dmix, it might not be able to mix two different samplerates (Spotify is 44.1k, Brave outputs 48k), and so far I haven't been able to make a configuration that would work...

**Comment by bondagit** [comment]
dmix performs resampling in case the playback samplerate doesn't match the plug samplerate.
So my suggestion is to configure the two Sources @48KHz and the two dmix plugs @48KHz.
dmix will handle resampling to reach the plug configured samplerate in case this is required. 
You can try to simulate the behaviour using _aplay_ on a sample file. With _aplay_ use the option _-vv_ to view the audio chain used.

**Comment by whyronimus** [comment]
Is it necessary to provide ALSA configuration to work with JACK audio? 

The plughw:RAVENNA device is recognized in JACK (no ALSA config currently) and can be activated. My problem is that audio is received in JACK Clients, but when I try to send via JACK no audio is going out apparently. I tried to route the output signal via Dante Controller, but the Icons stay at "prohibited" stage.

**Comment by bondagit** [comment]
To work with Jack make sure you the daemon parameter "tic_frame_size_at_1fs" set to 64.
See discussion at #38

**Comment by Matzeee** [comment]
@bondagit I tried your solution, but I get overrruns permanently.
If anyone wondering, there is a typo in the binding for sink_2: `bindings { 0 2 1 3 }` instead `bindings { 2 0 3 1 } ` works.
```
sudo arecord -D plug:sink_2 -c 2 -f S24_3LE -r 48000 -t wav sink.wav
Recording WAVE 'sink.wav' : Signed 24 bit Little Endian in 3bytes, Rate 48000 Hz, Stereo
overrun!!! (at least 0.015 ms long)
overrun!!! (at least 0.015 ms long)
overrun!!! (at least 0.015 ms long)
^CAborted by signal Interrupt...
arecord: pcm_read:2221: read error: Interrupted system call
```

My goal is to use multiple stereo channels from aes67 with [liquidsoap](https://github.com/savonet/liquidsoap).
`main.liq`:
```
settings.frame.audio.samplerate := 48000

stream = input.alsa(device="plug:sink_1", fallible=false)
```
```
2023/09/28 12:50:42 [input.alsa:3] Overrun!
2023/09/28 12:50:42 [input.alsa:2] Trying to recover..
2023/09/28 12:50:42 [input.alsa:3] Overrun!
2023/09/28 12:50:42 [input.alsa:2] Trying to recover..
2023/09/28 12:50:44 [input.alsa:3] Overrun!
2023/09/28 12:50:44 [input.alsa:2] Trying to recover..
2023/09/28 12:50:44 [input.alsa:3] Overrun!
2023/09/28 12:50:44 [input.alsa:2] Trying to recover..
```


Using it directly works without problems, but it only picks the first 2 channels:
```
settings.frame.audio.samplerate := 48000

stream = input.alsa(device="hw:RAVENNA,0", fallible=false)
```

Do you have any idea how to get rid of the overruns (hopefully without using jack)?

**Comment by bondagit** [comment]
You have to configure a buffer on your audio chain with option _buffer_size_ in the sink definition.
For example:
```
pcm.sink_1 {
    type dsnoop
    ipc_key 0x11111
    slave {
        pcm "hw:CARD=RAVENNA"
        channels 4
        rate 48000
        format S24_3LE
        buffer_size 4096
    }
    bindings { 0 0 1 1 }
}
```
With _arecord_ you can use the option -vv to check the audio chain and the volume meter.



> If anyone wondering, there is a typo in the binding for sink_2: `bindings { 0 2 1 3 }` instead `bindings { 2 0 3 1 } ` works.

you are right on this.

**Comment by Matzeee** [comment]
Works perfectly.
Thank you very much, also for the whole project and providing documentation on installation and the setup with Dante-devices!

By adding the buffer, this similar approach also works: https://github.com/savonet/liquidsoap/issues/293#issuecomment-183895230
Can you imagine any advantages or disadvantages? Probably just a matter of taste?

**Comment by bondagit** [comment]
> Can you imagine any advantages or disadvantages? Probably just a matter of taste?

The bigger the buffer the higher the latency. Of course, if you are recording to files this is irrelevant.

**Comment by Matzeee** [comment]
Thank you again!

Actually, I meant if there is any difference between

```
pcm.sink_1 {
    type dsnoop
    ipc_key 0x11111
    slave {
        pcm "hw:CARD=RAVENNA"
        channels 4
        rate 48000
        format S24_3LE
        buffer_size 4096
    }
    bindings { 0 0 1 1 }
}
```
`stream = input.alsa(device="plug:sink_1", fallible=false)`

and

```
liq.dsnoop1 {
  type dsnoop
  ipc_key 12345
  slave {
    #pcm "hw:RAVENNA,0"
    pcm "hw:CARD=RAVENNA"
    channels 4
    rate 48000
    format S24_3LE
    buffer_size 4096
  }
  bindings { 0 0 1 1 }
}

liq.input1 {
  type plug
  slave {
    pcm "liq.dsnoop1"
  }
}
```
`stream = input.alsa(device="liq.input1", fallible=false)`
But I assume, it is only cosmetically?


Another, maybe related question: We are running this setup now for a week and at least two times I heard very distorted audio for a few minutes, reminded me of how it sounded before adding `buffer_size 4096`.

Would you try to increase `buffer_size` or any other value of the daemon?

`Delay (samples)` in the sinks config is currently set to `576 - 12ms@48kHz`

In the global config we got:
```
Daemon	
bondagit-1.6.5

Audio Config
Safety Playout delay (samples)	
0
TIC frame size @1FS (samples)	
48 - 1ms
Max TIC frame size (samples)	
1024
Initial Sample rate	
48 kHz
```

**Comment by bondagit** [comment]
> We are running this setup now for a week and at least two times I heard very distorted audio for a few minutes

Try to check if this depends on CPU scaling events and disable them in case:
https://github.com/bondagit/aes67-linux-daemon#notes and https://github.com/bondagit/aes67-linux-daemon/issues/96

**Comment by bondagit** [comment]
> Do you recommend a separate sink / multicast flow per stereo pair or putting all channels in a single one?

it depends on your setup. My suggestion is to run some tests and check what fits better your needs,

**Comment by Matzeee** [comment]
> Try to check if this depends on CPU scaling events and disable them in case: https://github.com/bondagit/aes67-linux-daemon#notes and #96

I thought, i had done that already, but:
```
$ cat /proc/sys/kernel/perf_cpu_time_max_percent
25
$ sudo sysctl -w kernel.perf_cpu_time_max_percent=0
kernel.perf_cpu_time_max_percent = 0
$ cat /proc/sys/kernel/sched_rt_runtime_us
950000
$ sudo sysctl -w kernel.sched_rt_runtime_us=1000000
kernel.sched_rt_runtime_us = 1000000
$ cat /sys/devices/system/cpu/cpu0/cpufreq/scaling_governor
schedutil
```
Is it necessary to apply those settings after rebooting or after each kernelupdate?

**Comment by bondagit** [comment]
I would follow the recommandations at 5.3 of:

https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/managing_monitoring_and_updating_the_kernel/configuring-kernel-parameters-at-runtime_managing-monitoring-and-updating-the-kernel

**Comment by Matzeee** [comment]
Following your link, I found out, `sysctl -w kernel…` only lasts until rebooting and added:
```
cat /etc/sysctl.d/custom-aes67.conf
kernel.perf_cpu_time_max_percent = 0
kernel.sched_rt_runtime_us = 1000000
```
It seemed to help, there were no scaling events in `/var/log/kern.log`, but after some hours, we had distortions again.
Next, I did (thanks to your hint in the readme):
```
echo performance | sudo tee /sys/devices/system/cpu/cpu0/cpufreq/scaling_governor
echo performance | sudo tee /sys/devices/system/cpu/cpu1/cpufreq/scaling_governor
echo performance | sudo tee /sys/devices/system/cpu/cpu2/cpufreq/scaling_governor
echo performance | sudo tee /sys/devices/system/cpu/cpu3/cpufreq/scaling_governor
```
Since then, its been fine. But I did not find a nice way to make this permanent yet (besides using cron).

Our next try would be another machine or an ethernet adapter with hardware PTPv2 / IEEE 1588v2 support.
E.g. one with this chipset: https://www.asix.com.tw/en/product/USBEthernet/Super-Speed_USB_Ethernet/AX88179A
Would the driver / daemon make use of it?

And it turned out, it is necessary to recompile and reinstall after Ubuntu came up with a kernelupdate.
